runtime/tests/checker/transactions_test.go

```diff
@@ -398,82 +727,292 @@ func TestCheckInvalidTransactionSelfMoveToFunction(t *testing.T) {
+                      prepare() {
+                          let txs = [self]
+                      }
+                  }
+              }
+            `)
+
+			errs := RequireCheckerErrors(t, err, 1)
+
+			assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+		})
+
+		t.Run("into dictionary literal", func(t *testing.T) {
+
+			t.Parallel()
+
+			_, err := ParseAndCheck(t, `
+
+              transaction {
+
+                  role buyer {
+                      prepare() {
+                          let txs = {"self": self}
+                      }
+                  }
+              }
+            `)
+
+			errs := RequireCheckerErrors(t, err, 1)
+
+			assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+		})
+
+		t.Run("into function", func(t *testing.T) {
+
+			t.Parallel()
+
+			_, err := ParseAndCheck(t, `
+
+              fun test(_ x: AnyStruct) {}
+
+              transaction {
+
+                  role buyer {
+                      prepare() {
+                          test(self)
+                      }
+                  }
+              }
+            `)
+
+			errs := RequireCheckerErrors(t, err, 1)
+
+			assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+		})
+	})
+
+	t.Run("indirect", func(t *testing.T) {
+		t.Parallel()
+
+		t.Run("variable declaration", func(t *testing.T) {
+
+			t.Parallel()
+
+			_, err := ParseAndCheck(t, `
+
+              transaction {
+
+                  role buyer {}
+
+                  execute {
+                      let x = self.buyer
+                  }
+              }
+            `)
+
+			errs := RequireCheckerErrors(t, err, 1)
+
+			assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+		})
+
+		t.Run("into array literal", func(t *testing.T) {
+
+			t.Parallel()
+
+			_, err := ParseAndCheck(t, `
+
+              transaction {
+
+                  role buyer {}
+
+                  execute {
+                      let roles = [self.buyer]
+                  }
+              }
+            `)
+
+			errs := RequireCheckerErrors(t, err, 1)
+
+			assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+		})
+
+		t.Run("into dictionary literal", func(t *testing.T) {
+
+			t.Parallel()
+
+			_, err := ParseAndCheck(t, `
+
+              transaction {
+
+                  role buyer {}
+
+                  execute {
+                      let roles = {"buyer": self.buyer}
+                  }
+              }
+            `)
+
+			errs := RequireCheckerErrors(t, err, 1)
+
+			assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+		})
+
+		t.Run("into function", func(t *testing.T) {
+
+			t.Parallel()
+
+			_, err := ParseAndCheck(t, `
+
+              fun test(_ x: AnyStruct) {}
+
+              transaction {
+
+                  role buyer {}
+
+                  execute {
+                      test(self.buyer)
+                  }
+              }
+            `)
+
+			errs := RequireCheckerErrors(t, err, 1)
+
+			assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+		})
+
+	})
```

Comment:
I was trying more ways to break the type checking, and seems the following does not give any checker errors:
```cadence
fun test(_ x: AnyStruct) {}

transaction {

    role buyer {}

    execute {
        test(&self.buyer as &AnyStruct)  // role inside a casting expression
    }
}
```

While it may be harmless, I think we should also reject such cases (thinks it's the same reason as https://github.com/onflow/cadence/pull/2257#discussion_r1083138732).

Reply:
We could, but we currently don't reject references to non-first class values anywhere else.

References to roles are not useful, but why would we reject it?

Reply:
> We could, but we currently don't reject references to non-first class values anywhere else.
>
> References to roles are not useful, but why would we reject it?

I think 'not useful / not being able to use anywhere' is a good enough reason to reject it. By statically rejecting, a developer would know that they are writing some wrong code, as early as possible.

Having said that, I also just realized, whether we are rejecting it or not, may also depend on how we 'explain' roles to users: If we are going to say a role is a 'value' inside a transaction, being able to take references is OK I guess. But if we are pitching roles as just a named-scope, then taking a reference to a scope sounds odd (I'm looking at roles from this second perspective so far, hence the suggestion). I mean, we may implement it as a value under the hood, but just wanted to say that the validations may depend on the story we want to tell to the users.

Reply:
anyway, it shouldn't be a blocker for the PR

Reply:
@SupunS Great points! It's easy to add, so lets do it. I'll make the check recursive, unless you think it should not be.

Reply:
should be on creating references?
i.e: roles can only be used in member access expressions - to access its members?

Reply:
Sounds good, I'll add that in a follow-up PR, this PR is already large and there are already two PRs stacked on this one

Reply:
Opened #2268